core/lib/dal/src/connection/mod.rs

```diff
@@ -56,46 +55,80 @@ impl ConnectionPoolBuilder {
             DbVariant::Master => get_master_database_url()?,
             DbVariant::Replica => get_replica_database_url()?,
             DbVariant::Prover => get_prover_database_url()?,
+            DbVariant::TestTmp => create_test_db().await.to_string(),
         };
-        Ok(self.build_inner(&database_url).await)
+        self.build_inner(&database_url)
+            .await
+            .context("build_inner()")
     }
 
-    pub async fn build_inner(&self, database_url: &str) -> ConnectionPool {
+    pub async fn build_inner(&self, database_url: &str) -> anyhow::Result<ConnectionPool> {
         let max_connections = self
             .max_size
             .unwrap_or_else(|| parse_env("DATABASE_POOL_SIZE"));
 
         let options = PgPoolOptions::new().max_connections(max_connections);
-        let mut connect_options: PgConnectOptions = database_url.parse().unwrap_or_else(|err| {
-            panic!("Failed parsing {:?} database URL: {}", self.db, err);
-        });
+        let mut connect_options: PgConnectOptions = database_url
+            .parse()
+            .with_context(|| format!("Failed parsing {:?} database URL", self.db))?;
         if let Some(timeout) = self.statement_timeout {
             let timeout_string = format!("{}s", timeout.as_secs());
             connect_options = connect_options.options([("statement_timeout", timeout_string)]);
         }
         let pool = options
             .connect_with(connect_options)
             .await
-            .unwrap_or_else(|err| {
-                panic!("Failed connecting to {:?} database: {}", self.db, err);
-            });
+            .with_context(|| format!("Failed connecting to {:?} database", self.db))?;
         tracing::info!(
             "Created pool for {db:?} database with {max_connections} max connections \
              and {statement_timeout:?} statement timeout",
             db = self.db,
             statement_timeout = self.statement_timeout
         );
-        ConnectionPool::Real(pool)
+        Ok(ConnectionPool(pool))
     }
 }
 
-#[derive(Debug, Clone)]
-pub enum ConnectionPool {
-    Real(PgPool),
-    Test(TestPool),
+const PREFIX: &str = "test-";
+
+pub(super) async fn create_test_db() -> url::Url {
+    use rand::Rng as _;
+    use sqlx::{Connection as _, Executor as _};
+    let db_url = crate::get_test_database_url().unwrap();
+    let mut db_url = url::Url::parse(&db_url).unwrap();
+    let db_name = db_url.path()[1..].to_string();
+    let db_copy_name = format!("{PREFIX}{}", rand::thread_rng().gen::<u64>());
+    db_url.set_path("");
+    let mut conn = loop {
+        if let Ok(conn) = sqlx::PgConnection::connect(db_url.as_ref()).await {
+            break conn;
+        } else {
+            tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+        }
+    };
```

Comment:
Nit: It could make sense to limit the number of attempts and panic with the last error message when they are exhausted. Then again, depends on whether we assume that test DB URL is always valid.

Reply:
done